flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/agglomerativeclustering/AgglomerativeClustering.java

```diff
@@ -0,0 +1,415 @@
+                AgglomerativeClusteringParams<AgglomerativeClustering> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public AgglomerativeClustering() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        Integer k = getK();
+        Double distanceThreshold = getDistanceThreshold();
+        Preconditions.checkArgument(
+                (k == null && distanceThreshold != null)
+                        || (k != null && distanceThreshold == null),
+                "One of param k and distanceThreshold should be null.");
+
+        if (getLinkage().equals(LINKAGE_WARD)) {
+            String distanceMeasure = getDistanceMeasure();
+            Preconditions.checkArgument(
+                    distanceMeasure.equals(EuclideanDistanceMeasure.NAME),
+                    distanceMeasure
+                            + " was provided as distance measure while linkage was ward. Ward only works with euclidean.");
+        }
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Row> dataStream = tEnv.toDataStream(inputs[0]);
+
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), Types.INT),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getPredictionCol()));
+
+        OutputTag<Tuple4<Integer, Integer, Double, Integer>> mergeInfoOutputTag =
+                new OutputTag<Tuple4<Integer, Integer, Double, Integer>>("MERGE_INFO") {};
+
+        SingleOutputStreamOperator<Row> output =
+                dataStream.transform(
+                        "doLocalAgglomerativeClustering",
+                        outputTypeInfo,
+                        new LocalAgglomerativeClusteringOperator(
+                                getFeaturesCol(),
+                                getLinkage(),
+                                getDistanceMeasure(),
+                                getK(),
+                                getDistanceThreshold(),
+                                getComputeFullTree(),
+                                mergeInfoOutputTag));
+        output.getTransformation().setParallelism(1);
+
+        Table outputTable = tEnv.fromDataStream(output);
+
+        DataStream<Tuple4<Integer, Integer, Double, Integer>> mergeInfo =
+                output.getSideOutput(mergeInfoOutputTag);
+        Table mergeInfoTable =
+                tEnv.fromDataStream(mergeInfo)
+                        .as("clusterId1", "clusterId2", "distance", "sizeOfMergedCluster");
+
+        return new Table[] {outputTable, mergeInfoTable};
```

Comment:
Let's describe in JavaDoc the column names of the `mergeInfoTable` and their meanings.

Besides, do you think it would be better to add some method like `Table getMergeInfo()`, instead of making the merge info table the second output table? I'm not sure the current API corresponds to the syntax of Flink ML AlgoOperators. The current API also makes it difficult to connect `AgglomerativeClustering` stage with other stages in a `Pipeline`.

Reply:
Thanks for the comment. I have added the following description in Java doc.
```
The output contains two tables. The first one assigns one cluster Id for each data point.
The second one contains the information of merging two clusters at each step. The data format
of the merging information is (clusterId1, clusterId2, distance, sizeOfMergedCluster).
```

I am not sure about adding a `Table getMergeInfo()` here, since it is somehow adding a public API. 

For the semantic of  `AgglomerativeClustering`, I made it an AlgoOperator because (1) it is not a Estimator since it has not model data and cannot be used to predict some new data. (2) It is also not a Transformer. 

It is clear that not all AlgoOperator/Estimator could be used in Pipeline. For those multi-input/output stages, we have `Graph`.

Reply:
Thanks for the explanation. I agree with it that the current API is better.